### ISSUE TITLE
Failed to get document because the client is offline. 

### ISSUE THREAD

**Issue by UD-UD** [EddyVerbruggen/nativescript-plugin-firebase#631]
I am trying to develop an android app using nativescript and firebase as backend support.To use Firebase I am using this plugin. I am trying to use Firestore for my app .I get this error each time I try to start the app.When I used realtime database ,I was easily able to access all the data, so this is not a network or configuration issue. Below is the code I am using to connet firebase and use firestore.
```js
 firebase.initializeApp({}).then((data)=>{
           console.log(data) //it prints the databaseURL and it is correct.
        });
        var db = firebase.firestore();
        db.collection("users").doc('<docKey>').get().then((doc) => {
            if (doc.exists) {
                this.data =doc.data();
            } else{
                this.data ="No such document!";
            }
        }).catch((error) => {
            console.log(error);
        });
``` 
### Error : com.google.firebase.firestore.FirebaseFirestoreException: Failed to get document because the client is offline.

**Comment by EddyVerbruggen** [comment]
Did you try Googling it? Please do and provide some pointers here if you have time: https://www.google.com/search?q=com.google.firebase.firestore.FirebaseFirestoreException%3A+Failed+to+get+document+because+the+client+is+offline.&oq=com.google.firebase.firestore.FirebaseFirestoreException%3A+Failed+to+get+document+because+the+client+is+offline

**Comment by UD-UD** [comment]
Yes Eddy, I did some initial research but didn't find any valuable solution. Below are some links where this same issue is raised.

[ No solution provided](https://forums.expo.io/t/firebaseerror-failed-to-get-document-because-the-client-is-offline/3982)

[ Still waiting for some solution ](https://stackoverflow.com/questions/46620957/com-google-firebase-firestore-firebasefirestoreexception-failed-to-get-document)

[ Here it is said a fix was rolled out but still the issue is faced by some users](https://stackoverflow.com/questions/46602889/firebase-cloud-firestore-throws-client-is-offline)

I guess this is happening for Android App development only as most of the above issue was raised while developing for android.

Solution to this issue will be really a great help.

Thanks.

**Comment by EddyVerbruggen** [comment]
So do you think this is an issue with the plugin, or the Firebase Android SDK?

**Comment by UD-UD** [comment]
I am also in the same dilemma. I have raised a issue in firebase support also. Let me get in touch with them and discuss the issue and will update here accordingly. But in the meantime it would be great if you provide some insights regarding this or workaround.

**Comment by ananddayalan** [comment]
If you have changed the project id but not changed in firebase.initializeApp, you'll get this same error. Please check and see if that's the case.

**Comment by Arctic45** [comment]
I got this problem on Android. It happens after reconnecting to the internet, even though any other network calls work fine.

**Comment by jacopo69** [comment]
I solved it after modifing security rules on that document

**Comment by EddyVerbruggen** [comment]
Can you share what you changes exactly?

**Comment by jacopo69** [comment]
Simply check if you have permissions to access your document. In my case I forgot to set the read permission:
```
match /channels/{channel} {
     allow read: if request.auth.uid != null;
    }
```

**Comment by Kravet** [comment]
I had the same error, when I used wrong path to the document. Anyway the error message is misleading.

**Comment by shuvojit007** [comment]
I solved this after using addOnSuccessListener instead of OnCompleteListener.

OnCompleteListener is always called if completed successfully, or there was an error.  So we should check the task for success or failure. Or we can split it up into an onSuccess and onFailure callbacks that get invoked depending on the status of the task.

**Comment by mudi256** [comment]
i'm facing this issue on kitkat when adding inApp messaging dependency

**Comment by Robotecom** [comment]
> I solved this issue after using addOnSuccessListener instead of OnCompleteListener. OnCompleteListener is always called if the task was completed successfully, or there was an error. So we should check the task for success or failure. Or we can split it up into an onSuccess and onFailure callbacks that get invoked depending on the status of the task.

This issue seems to be related to the addOnCompleteListener only ,I also had this error ,and eventually resolved after I used the onSuccess and onFailure listeners instead

**Comment by DeanPack** [comment]
I'm facing this same issue right now. I'm not use oncomplete listener. I'm just using a straight up call to the database. This only happens sometimes for some users, and seems to fix itself when they login and log out. Here is the call to the database that is erroring out. Has anyone figured this out?
DocumentSnapshot userLivestream = await Firestore.instance.collection('livestreams').document(userId).get();

**Comment by manojdcoder** [comment]
Did you try the above workaround suggested by others, using `addOnSuccessListener` instead of `OnCompleteListener`.

**Comment by Meghraj-coder** [comment]
> If you have changed the project id but not changed in firebase.initializeApp, you'll get this same error. Please check and see if that's the case.

how to know where does the project id lie ?

**Comment by elis** [comment]
> Did you try Googling it? Please do and provide some pointers here if you have time: https://www.google.com/search?q=com.google.firebase.firestore.FirebaseFirestoreException%3A+Failed+to+get+document+because+the+client+is+offline.&oq=com.google.firebase.firestore.FirebaseFirestoreException%3A+Failed+to+get+document+because+the+client+is+offline

Hey Eddy, guess what, your snarky comment is at the top of google searches.

**Comment by EddyVerbruggen** [comment]
@elis What's so snarky about asking folks to help me figuring out the cause of an issue? Do you have any idea how hard it is to triage issues like this one?

**Comment by subhadippramanik** [comment]
@UD-UD How did you solve this issue? I am facing same error and nothing from Google could solve it. Looking for your help if possible.

**Comment by saulmaos** [comment]
Same error on android using kotlin. My app works well but a user is facing that bug. I'm using success and failure listeners. The user has the same app version than me, everything is the same. I can login with his account and the app works but on his phone he cannot load any data. He is able to login successfully by the way. Anyway, what does "the client is offline" mean? does it mean that firestore server is down? if so, why i'm able to load data but my user cannot? he's facing the problem since one week ago...

**Comment by SujitSingh** [comment]
This is a very weird issue. My web application is throwing the reported error in the browser's console but the Firebase API is fetching data most of the time(not always). Whenever it is not, it throws the reported error message.
Used rules in my app - 
```JSON
{
  "rules": {
    ".read": true,
    ".write": true,
  }
}
```
The same error happened yesterday as well but disappeared after a few seconds and started working properly.

**Comment by Joshfindit** [comment]
Insight from my experience today might help here:

I got `Failed to get document because the client is offline.` when I did two things that might get flagged by Google's firewall:

- Repeated requests for IDs that don't exist
  (happened during prototyping due to a logic error in my code)
- Too many requests at once
  (possibly 11000, but I didn't investigate. Even then I got 4 results back before the error on the rest of the requests)

_Note: both of those are authenticated requests on a database with rules set for public read and authorized write._

**Comment by vhmolinar** [comment]
That error might occur whenever device is offline and there is no data inside firestore local cache.
So, try to wrap get calls in a try/catch, handle the failure normally and you should be fine.
